frontend/src/routes/manage/Realm/Content/Edit/EditMode/index.tsx

```diff
@@ -0,0 +1,174 @@
+        onCompleted?: () => void,
+        onError?: (error: Error) => void,
+    ) => void;
+};
+
+export const EditMode: React.FC<EditModeProps> = ({
+    realm: realmRef,
+    index,
+    onSave,
+    onCancel,
+    onCompleted,
+    onError,
+}) => {
+    const { t } = useTranslation();
+
+    const { id: realmId, blocks } = useFragment(graphql`
+        fragment EditModeRealmData on Realm {
+            id
+            blocks {
+                id
+                title
+                __typename
+                ... on TextBlock { ... TextEditModeBlockData }
+                ... on SeriesBlock { ... SeriesEditModeBlockData }
+            }
+        }
+    `, realmRef);
+    const block = blocks[index];
+    const { id, title, __typename: type } = block;
+
+
+    const form = useForm<EditModeFormData>({
+        defaultValues: {
+            type: type as "TextBlock" | "SeriesBlock",
+        },
+    });
+    const editModeRef = useRef<EditModeRef>(null);
+
+    const onSubmit = form.handleSubmit(data => {
+        // Empty titles should set the field to `null`
+        // This is to avoid empty headings when rendering the block.
+        // In the future we might want to check this at render site,
+        // and we also might want more sophisticated checks
+        // (like "all whitespace").
+        const processData = { ...data, title: data.title || null };
+
+        const currentEditModeRef = editModeRef.current ?? bug("`editModeRef` not bound");
+
+        onSave?.();
+
+        if (id.startsWith("cl")) {
+            currentEditModeRef.create(
+                realmId,
+                index,
+                processData,
+                onCompleted,
+                onError,
+            );
+        } else {
+            currentEditModeRef.save(
+                id,
+                processData,
+                onCompleted,
+                onError,
+            );
+        }
+    });
+
+
+    return <FormProvider<EditModeFormData> {...form}>
+        <form onSubmit={onSubmit}>
+            <EditModeButtons onCancel={onCancel} />
+            <h2 css={{ margin: "16px 0" }}>
+                <Input
+                    css={{ display: "block" }}
+                    placeholder={t("manage.realm.content.title")}
+                    defaultValue={title ?? ""}
+                    {...form.register("title")}
+                />
+            </h2>
+            {match(block.__typename, {
+                "TextBlock": () => <EditTextBlock ref={editModeRef} block={block} />,
+                "SeriesBlock": () => <EditSeriesBlock ref={editModeRef} block={block} />,
+            })}
+        </form>
+    </FormProvider>;
+};
+
+
+type EditModeButtonsProps = {
+    onCancel?: () => void;
+};
+
+const EditModeButtons: React.FC<EditModeButtonsProps> = ({ onCancel }) => {
+    const { t } = useTranslation();
+
+    const modalRef = useRef<ConfirmationModalHandle>(null);
+
+    return <ButtonGroup css={{ marginTop: -24 }}>
```

Comment:
Magic number? I guess absolute positioning and `top: 0` is more idiomatic?

Reply:
Nop:

![image](https://user-images.githubusercontent.com/123272/150397106-754aeb1c-9494-4c59-bdb7-35ff696ca5b4.png)

Also it's not the only magic number in our CSS. :wink:

Reply:
Ye sure, css stuff is full of magic numers. And by `top:0` I of course also meant `right: 0`. But yeah, nevermind.